src/main/java/org/spongepowered/server/launch/VanillaServerMain.java

```diff
@@ -138,63 +145,137 @@ public static void main(String[] args) throws Exception {
     private static boolean downloadMinecraft(Path base, boolean autoDownload) throws IOException, NoSuchAlgorithmException {
         // Make sure the Minecraft server is available, or download it otherwise
         Path path = base.resolve(MINECRAFT_SERVER_LOCAL);
-        if (Files.notExists(path) && (!autoDownload || !downloadVerified(MINECRAFT_SERVER_REMOTE, path))) {
-            return false;
+        if (Files.notExists(path)) {
+            if (!autoDownload) {
+                return false;
+            }
+
+            System.out.println("Downloading the versions manifest...");
+
+            // Download the file with all of the Minecraft versions information
+            JsonValue versions = downloadJson(MINECRAFT_MANIFEST_REMOTE);
+
+            String versionManifestRemote = null;
+
+            // Find the current version manifest URL
+            for (JsonValue versionInfo : versions.asObject().get("versions").asArray()) {
+                JsonObject obj = versionInfo.asObject();
+
+                String versionId = obj.get("id").asString();
+                if (versionId.equals(MINECRAFT_SERVER_VERSION)) {
+                    versionManifestRemote = obj.get("url").asString();
+                    break;
+                }
+            }
+
+            if (versionManifestRemote == null) {
+                throw new RuntimeException("Could not find " + MINECRAFT_SERVER_VERSION + "'s manifest URL");
+            }
+
+            JsonValue versionManifest = downloadJson(versionManifestRemote);
+            JsonObject serverObj = versionManifest.asObject()
+                    .get("downloads").asObject()
+                    .get("server").asObject();
+
+            // Find the server URL and SHA-1 digest
+            String serverRemote = serverObj.get("url").asString();
+            String sha1 = serverObj.get("sha1").asString();
+
+            downloadAndVerify(serverRemote, path, sha1);
         }
 
-        // Make sure Launchwrapper is available, or download it otherwise
         path = base.resolve(LAUNCHWRAPPER_LOCAL);
-        return Files.exists(path) || (autoDownload && downloadVerified(LAUNCHWRAPPER_REMOTE, path));
+
+        if (!Files.exists(path)) {
+            if (!autoDownload) {
+                return false;
+            }
+
+            // Make sure Launchwrapper is available, or download it otherwise
+            download(LAUNCHWRAPPER_REMOTE, path);
+        }
+
+        return true;
     }
 
+    private static JsonValue downloadJson(String remote) throws IOException {
+        URLConnection con = new URL(remote).openConnection();
```

Comment:
Since we're not doing anything with the connection, just call `url.openStream()`. Merge everything into one try statement (so the `InputStreamReader` is also closed properly):

```java
try (Reader reader = new InputStreamReader(new URL(remote).openStream(), StandardCharsets.UTF_8))) {
    return Json.parse(reader);
}
```

Also, according to the Javadocs in minimal-json, their reader is already buffered so there is no need to use a `BufferedInputStream` here.